### PR TITLE
run KMT with custom ssh connector

### PR DESCRIPTION
**Pull request by brycekahle**
### What does this PR do?

Runs KMT with a custom program to handle SSH between gitlab runner <-> metal instance <-> microvm.

### Motivation

This allows us to handle SSH options directly, and to report metrics on SSH connection stability.

### Additional Notes

Will update the test-infra-definitions version once https://github.com/DataDog/test-infra-definitions/pull/388 is merged.

https://datadoghq.atlassian.net/browse/EBPF-237

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by usamasaqib — APPROVED** [review]

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 1]
```diff
@@ -338,7 +362,7 @@ kernel_matrix_testing_run_tests_x64:
 kernel_matrix_testing_run_tests_arm64:
   extends:
     - .kernel_matrix_testing_run_tests
-  needs: ["kernel_matrix_testing_setup_env", "upload_dependencies_arm64", "upload_system_probe_tests_arm64", "upload_minimized_btfs_arm64"]
+  needs: ["kernel_matrix_testing_setup_env", "upload_dependencies_arm64", "upload_system_probe_tests_x64", "upload_system_probe_tests_arm64", "upload_minimized_btfs_arm64"]
```
Comment (chouetz) [inline .gitlab/kernel_version_testing/system_probe.yml:365]: ❓ why does this `arm64` step would need a dependency on an `x64` job?

**Review by chouetz — APPROVED** [review]
If the modification done on the line I commented was on purpose it's ok for me

File: .gitlab/kernel_version_testing/system_probe.yml (hunk 1 above)
Comment (brycekahle) [reply .gitlab/kernel_version_testing/system_probe.yml]: Good catch, fixed. I tried to find out where that came in, but because I squashed I lost that history.

Comment (brycekahle) [reply .gitlab/kernel_version_testing/system_probe.yml:365]: Turns out this was required. It is because the e2e runner image is only `amd64`, so we need the `connector` program built for that architecture regardless if the metal+microvms will be `arm64`.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -166,7 +166,8 @@ variables:
   DATADOG_AGENT_NIKOS_BUILDIMAGES: v20878799-a2f77ae
   DATADOG_AGENT_BTF_GEN_BUILDIMAGES_SUFFIX: ""
   DATADOG_AGENT_BTF_GEN_BUILDIMAGES: v20878799-a2f77ae
-  TEST_INFRA_DEFINITIONS_BUILDIMAGES: 40c538510ddf
+  TEST_INFRA_DEFINITIONS_BUILDIMAGES_SUFFIX: "-dev"
```
Comment (vboulineau) [inline .gitlab-ci.yml:169]: Should be removed before merging on `main`

File: test/new-e2e/system-probe/connector/sshtools/communicator.go [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+// easy to implement by custom loggers, if they don't do so already anyway.
+type Logger interface {
+	Println(v ...interface{})
+}
+
+// Communicator allows for executing commands on a remote host over SSH, it is
+// not thread safe. New communicator is not connected by default, however,
+// calling Start or Upload on not connected communicator would try to establish
+// SSH connection before executing.
+type Communicator struct {
```
Comment (vboulineau) [inline test/new-e2e/system-probe/connector/sshtools/communicator.go:30]: It seems to me like this code should be common code and integrate or replace code in `test/new-e2e/pkg/utils/clients/ssh.go`

**Review by vboulineau — CHANGES_REQUESTED** [review]

File: test/new-e2e/system-probe/connector/sshtools/communicator.go (hunk 3 above)
Comment (brycekahle) [reply test/new-e2e/system-probe/connector/sshtools/communicator.go:30]: It could be, but it would change the API of those functions quite a bit. Can we leave that for a future PR?

File: .gitlab-ci.yml (hunk 2 above)
Comment (brycekahle) [reply .gitlab-ci.yml:169]: > Will update the test-infra-definitions version once https://github.com/DataDog/test-infra-definitions/pull/388 is merged.

I did mention that in the PR description

File: pkg/network/protocols/testutil/serverutils.go [hunk 4]
```diff
@@ -35,7 +35,7 @@ func RunDockerServer(t testing.TB, serverName, dockerPath string, env []string,
 	ctx, cancel := context.WithCancel(context.Background())
 	t.Cleanup(cancel)
 
-	cmd := exec.CommandContext(ctx, "docker-compose", "-f", dockerPath, "up")
+	cmd := exec.CommandContext(ctx, "docker-compose", "-f", dockerPath, "up", "--force-recreate", "--remove-orphans")
```
Comment (guyarb) [inline pkg/network/protocols/testutil/serverutils.go:38]: why did you add `force-recreate`?

File: .gitlab/kernel_version_testing/system_probe.yml (hunk 1 above)
Comment (chouetz) [reply .gitlab/kernel_version_testing/system_probe.yml]: Could it deserve a comment?

File: pkg/network/protocols/testutil/serverutils.go (hunk 4 above)
Comment (brycekahle) [reply pkg/network/protocols/testutil/serverutils.go:38]: I was getting errors about the network `testdata-network-1` not being present. It seemed like some sort of race with the cleanup.

**Review by guyarb — APPROVED** [review]

File: .gitlab-ci.yml (hunk 2 above)
Comment (brycekahle) [reply .gitlab-ci.yml]: Updated

**Review by vboulineau — APPROVED** [review]

File: .gitlab/kernel_version_testing/system_probe.yml (hunk 1 above)
Comment (brycekahle) [reply .gitlab/kernel_version_testing/system_probe.yml]: I fixed it, so no comment needed. 😄
